File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/main/CloudAccountsPresenter.java
package app.editors.manager.mvp.presenters.main;

import android.content.Context;
import android.util.Log;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import app.editors.manager.app.App;
import app.editors.manager.app.WebDavApi;
import app.editors.manager.managers.tools.AccountSqlTool;
import app.editors.manager.managers.tools.PreferenceTool;
import app.editors.manager.managers.tools.RetrofitTool;
import app.editors.manager.mvp.models.account.AccountsSqlData;
import app.editors.manager.mvp.models.response.ResponseUser;
import app.editors.manager.mvp.models.user.User;
import app.editors.manager.mvp.presenters.login.BaseLoginPresenter;
import app.editors.manager.mvp.views.main.CloudAccountsView;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import moxy.InjectViewState;
import okhttp3.Credentials;
import okhttp3.ResponseBody;
import retrofit2.HttpException;
import retrofit2.Response;

@InjectViewState
public class CloudAccountsPresenter extends BaseLoginPresenter<CloudAccountsView, ResponseUser> {

    public static final String TAG = CloudAccountsPresenter.class.getSimpleName();

    @Inject
    protected Context mContext;
    @Inject
    protected AccountSqlTool mAccountSqlTool;
    @Inject
    protected PreferenceTool mPreferenceTool;

    @Nullable
    private AccountsSqlData mContextAccount;
    @Nullable
    private AccountsSqlData mClickedAccount;

    private int mContextPosition;

    private CompositeDisposable mDisposable = new CompositeDisposable();

    private List<AccountsSqlData> mSelectedItems;
    private boolean mIsSelectionMode;

    public CloudAccountsPresenter() {
        App.getApp().getAppComponent().inject(this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mDisposable.clear();
    }

    public void getAccounts() {
        List<AccountsSqlData> accounts = mAccountSqlTool.getAccounts();
        if (mIsSelectionMode) {
            checkSelected(accounts);
            getViewState().onSelectionMode();
        }
        if (!accounts.isEmpty()) {
            getViewState().onSetAccounts(accounts);
        }
        checkUpdate(accounts);
    }

    private void checkSelected(List<AccountsSqlData> accounts) {
        if (mSelectedItems != null) {
            for (AccountsSqlData selectedAccount : mSelectedItems) {
                for (AccountsSqlData account : accounts) {
                    if (selectedAccount.getId().equals(account.getId())) {
                        account.setSelection(true);
                    }
                }
            }
            mSelectedItems.clear();
            for (AccountsSqlData account : accounts) {
                if (account.isSelection()) {
                    mSelectedItems.add(account);
                }
            }
        }
    }

    public void checkLogin(AccountsSqlData account, int position) {
        mClickedAccount = account;
        if (mIsSelectionMode) {
            addSelectionItem(account, position);
        } else {
            if (account.isOnline()) {
                getViewState().onShowClouds();
            } else if (account.isWebDav()) {
                loginWebDav(account);
            } else {
                login(account);
            }
        }
    }

    private void addSelectionItem(AccountsSqlData account, int position) {
        if (mSelectedItems.contains(account)) {
            account.setSelection(false);
            mSelectedItems.remove(account);
        } else {
            account.setSelection(true);
            mSelectedItems.add(account);
        }
        getViewState().onActionBarTitle(String.valueOf(mSelectedItems.size()));
        getViewState().onSelectedItem(position);
    }

    public void contextClick(AccountsSqlData account, int position) {
        mContextAccount = account;
        mContextPosition = position;
        getViewState().onShowBottomDialog(account);
    }

    private void loginWebDav(AccountsSqlData account) {
        if (account.getPassword() == null || account.getPassword().equals("")) {
            getViewState().onWebDavLogin(account);
            return;
        }
        final String credential = Credentials.basic(account.getLogin(), account.getPassword());
        final WebDavApi.Providers provider = WebDavApi.Providers.valueOf(account.getWebDavProvider());
        String path;
        if (provider == WebDavApi.Providers.OwnCloud || provider == WebDavApi.Providers.NextCloud ||
                provider == WebDavApi.Providers.WebDav) {
            path = account.getWebDavPath();
        } else {
            path = provider.getPath();
        }

        mDisposable.add(WebDavApi.getApi(account.getScheme() + account.getPortal())
                .capabilities(credential, path)
                .doOnSubscribe(disposable -> getViewState().onShowWaitingDialog())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(responseBody -> {
                    if (responseBody.isSuccessful() && responseBody.code() == 207) {
                        return responseBody.body();
                    } else {
                        throw new HttpException(responseBody);
                    }
                })
                .subscribe(body -> saveWebDav(account), this::fetchError));
    }

    private void saveWebDav(AccountsSqlData account) {
        setAccount(false);
        mPreferenceTool.setDefault();
        mPreferenceTool.setPortal(account.getPortal());
        mPreferenceTool.setLogin(account.getLogin());
        mPreferenceTool.setPassword(account.getPassword());
        mPreferenceTool.setVisitor(false);
        mPreferenceTool.setNoPortal(false);
        setAccount(true);
        getViewState().onAccountLogin();
    }

    private void login(AccountsSqlData account) {
        if (account.getToken() != null && !account.getToken().isEmpty()) {
            getViewState().onShowWaitingDialog();
            onGetToken(account.getToken(), account);
        } else {
            if (account.isWebDav()) {
                getViewState().onWebDavLogin(account);
            } else {
                mPreferenceTool.setScheme(account.getScheme());
                mPreferenceTool.setSslState(account.isSslState());
                mPreferenceTool.setSslCiphers(account.isSslCiphers());
                getViewState().onSignIn(account.getPortal(), account.getLogin());
            }
        }
    }

    @Override
    protected void onErrorUser(Response<ResponseUser> response) {
        if (mClickedAccount != null) {
            getViewState().onSignIn(mClickedAccount.getPortal(), mClickedAccount .getLogin());
        }
    }

    @Override
    protected void onGetUser(User user) {
        super.onGetUser(user);
        getViewState().onAccountLogin();
    }

    private void setAccount(boolean isOnline) {
        AccountsSqlData account = mAccountSqlTool.getAccount(mPreferenceTool.getPortal(), mPreferenceTool.getLogin(), mPreferenceTool.getSocialProvider());
        if (account != null) {
            account.setOnline(isOnline);
            mAccountSqlTool.setAccount(account);
        }
    }

    public void logout() {
        if (mContextAccount != null) {
            if (mContextAccount.isOnline()) {
                mPreferenceTool.setDefault();
            }
            mContextAccount.setOnline(false);
            mContextAccount.setPassword("");
            mContextAccount.setToken("");
            mAccountSqlTool.setAccount(mContextAccount);
            getViewState().onUpdateItem(mContextAccount);
        }
    }

    public void removeAccount() {
        if (mContextAccount != null) {
            if (mContextAccount.isOnline()) {
                mPreferenceTool.setDefault();
            }
            mAccountSqlTool.delete(mContextAccount);
            if (mAccountSqlTool.getAccounts().isEmpty()) {
                getViewState().onEmptyList();
            } else {
                getViewState().removeItem(mContextPosition);
            }
        }
    }

    public void signIn() {
        if (mContextAccount != null) {
            login(mContextAccount);
        }
    }

    public AccountsSqlData getContextAccount() {
        return mContextAccount;
    }

    public void longClick(AccountsSqlData account, int position) {
        mSelectedItems = new ArrayList<>();
        mIsSelectionMode = true;
        account.setSelection(true);
        mSelectedItems.add(account);
        getViewState().onSelectionMode();
        getViewState().onActionBarTitle(String.valueOf(mSelectedItems.size()));
    }

    public boolean onBackPressed() {
        if (mIsSelectionMode) {
            mIsSelectionMode = false;
            clearSelected();
            getViewState().onDefaultState();
            return true;
        }

        return false;
    }

    private void clearSelected() {
        for (AccountsSqlData account : mSelectedItems) {
            account.setSelection(false);
        }
        mSelectedItems.clear();
        mSelectedItems = null;
    }

    public void selectAll(List<AccountsSqlData> itemList) {
        for (AccountsSqlData account : itemList) {
            if (!account.isSelection() && !account.isIdNull()) {
                account.setSelection(true);
                mSelectedItems.add(account);
            }
        }
        getViewState().onActionBarTitle(String.valueOf(mSelectedItems.size()));
        getViewState().onNotifyItems();
    }

    public void deleteAll() {
        if (mSelectedItems != null) {
            for (AccountsSqlData account : mSelectedItems) {
                if (account.isOnline()) {
                    mPreferenceTool.setDefault();
                }
                mAccountSqlTool.delete(account);
            }
            onBackPressed();
            if (mAccountSqlTool.getAccounts().isEmpty()) {
                getViewState().onEmptyList();
            } else {
                getViewState().onSetAccounts(mAccountSqlTool.getAccounts());
            }
        }
    }

    public void deselectAll() {
        if (mSelectedItems != null) {
            for (AccountsSqlData account : mSelectedItems) {
                account.setSelection(false);
            }
            mSelectedItems.clear();
            getViewState().onActionBarTitle(String.valueOf(mSelectedItems.size()));
            getViewState().onNotifyItems();
        }
    }

    private void checkUpdate(List<AccountsSqlData> accounts) {
        for (AccountsSqlData account : accounts) {
            if (!account.isWebDav() && account.getToken() != null && !account.getToken().isEmpty()) {
                mDisposable.add(getUser(account)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .map(responseUser -> {
                            if (responseUser.getResponse() == null) {
                                return new User();
                            } else {
                                return responseUser.getResponse();
                            }
                        })
                        .subscribe(user -> updateAccount(user, account),
                                throwable -> Log.d(TAG, "checkUpdate: " + throwable.getMessage())));
            }
        }
    }

    private void updateAccount(User user, AccountsSqlData account) {
        if (!user.getDisplayNameHtml().isEmpty() && !account.getName().equals(user.getDisplayNameHtml())) {
            account.setName(user.getDisplayNameHtml());
        }
        if (!user.getAvatar().isEmpty() && !account.getAvatarUrl().equals(user.getAvatar())) {
            account.setAvatarUrl(user.getAvatar());
        }
        mAccountSqlTool.setAccount(account);
        getViewState().onNotifyItems();
    }

    private Observable<ResponseUser> getUser(AccountsSqlData account) {
        String token = account.getToken();
        RetrofitTool retrofitTool = new RetrofitTool(mContext);
        retrofitTool.setCiphers(account.isSslCiphers());
        retrofitTool.setSslOn(account.isSslState());
        return Observable.fromCallable(() -> {
            Response<ResponseUser> response = retrofitTool.getApi(account.getScheme() + account.getPortal())
                    .getUserInfo(token).execute();
            if (response.isSuccessful()) {
                return response.body();
            } else {
                return new ResponseUser();
            }
        });

    }

    @Override
    protected void onErrorHandle(ResponseBody responseBody, int responseCode) {
       // stub
    }

    public void restoreAccount() {
        final AccountsSqlData accountsSqlData = mAccountSqlTool.getAccountOnline();
        if (accountsSqlData != null) {
            mPreferenceTool.setScheme(accountsSqlData.getScheme());
            mPreferenceTool.setPortal(accountsSqlData.getPortal());
            mPreferenceTool.setToken(accountsSqlData.getToken());
            mPreferenceTool.setSslState(accountsSqlData.isSslState());
            mPreferenceTool.setSslCiphers(accountsSqlData.isSslCiphers());
        }
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/main/CloudAccountsPresenter.kt
package app.editors.manager.mvp.presenters.main

import android.accounts.Account
import app.documents.core.account.CloudAccount
import app.documents.core.login.LoginResponse
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.managers.exceptions.UrlSyntaxMistake
import app.editors.manager.mvp.models.account.AccountsSqlData
import app.editors.manager.mvp.models.response.ResponseUser
import app.editors.manager.mvp.models.user.User
import app.editors.manager.mvp.presenters.login.BaseLoginPresenter
import app.editors.manager.mvp.views.main.CloudAccountsView
import io.reactivex.disposables.Disposable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import lib.toolkit.base.managers.utils.AccountUtils
import moxy.InjectViewState
import retrofit2.Response
import java.util.*

//sealed class CloudAccountState {
//    class AccountLoadedState(val account: List<CloudAccount>): CloudAccountState()
//    object DefaultState : CloudAccountState()
//    object SelectedState : CloudAccountState()
//}

@InjectViewState
class CloudAccountsPresenter : BaseLoginPresenter<CloudAccountsView>() {

    companion object {
        val TAG: String = CloudAccountsPresenter::class.java.simpleName
    }

    init {
//        App.getApp().appComponent.inject(this)
    }

    private val service = App.getApp().appComponent.loginService

    var contextAccount: CloudAccount? = null
    private var mClickedAccount: CloudAccount? = null
    private var mSelectedItems: MutableList<CloudAccount>? = null
    private var isSelectionMode = false
    private var mContextPosition = 0
    private var disposable: Disposable? = null

    override fun onDestroy() {
        super.onDestroy()
        disposable?.dispose()
    }

    fun getAccounts() {
        CoroutineScope(Dispatchers.Default).launch {
            getTokens(accountDao.getAccounts())
        }
    }

    private suspend fun getTokens(accounts: List<CloudAccount>) {
        val accountsWithToken = accounts.map { account ->
            AccountUtils.getToken(
                context,
                Account(account.getAccountName(), context.getString(R.string.account_type))
            )?.let { token ->
                account.token = token
                return@map account
            } ?: run {
                return@map account
            }
        }
        withContext(Dispatchers.Main) {
//            viewState.onRender(CloudAccountState.AccountLoadedState(accountsWithToken))
        }
    }

    private fun checkSelected(accounts: List<CloudAccount>) {
        if (mSelectedItems != null) {
            for (selectedAccount in mSelectedItems!!) {
                for (account in accounts) {
                    if (selectedAccount.id == account.id) {
//                        account.isSelection = true
                    }
                }
            }
            mSelectedItems!!.clear()
            for (account in accounts) {
//                if (account.isSelection) {
//                    mSelectedItems!!.add(account)
//                }
            }
        }
    }

    fun checkLogin(account: CloudAccount, position: Int) {
        mClickedAccount = account
        if (isSelectionMode) {
            addSelectionItem(account, position)
        } else {
            if (account.isOnline) {
                viewState.onShowClouds()
            } else if (account.isWebDav) {
                loginWebDav(account)
            } else {
                login(account)
            }
        }
    }

    private fun addSelectionItem(account: CloudAccount, position: Int) {
        if (mSelectedItems!!.contains(account)) {
//            account.isSelection = false
            mSelectedItems!!.remove(account)
        } else {
//            account.isSelection = true
//            mSelectedItems!!.add(account)
        }
        viewState.onActionBarTitle(mSelectedItems!!.size.toString())
        viewState.onSelectedItem(position)
    }

    fun contextClick(account: CloudAccount?, position: Int) {
        contextAccount = account
        mContextPosition = position
        viewState.onShowBottomDialog(account)
    }

    private fun loginWebDav(account: CloudAccount) {
//        if (account.password == null || account.password == "") {
////            viewState.onWebDavLogin(account)
//            return
//        }
//        val credential = Credentials.basic(account.login, account.password)
//        val provider = WebDavApi.Providers.valueOf(account.webDavProvider)
//        val path: String
//        path =
//            if (provider == WebDavApi.Providers.OwnCloud || provider == WebDavApi.Providers.NextCloud || provider == WebDavApi.Providers.WebDav) {
//                account.webDavPath
//            } else {
//                provider.path
//            }
//        mDisposable.add(WebDavApi.getApi(account.scheme + account.portal)
//            .capabilities(credential, path)
//            .doOnSubscribe { disposable: Disposable? -> viewState.onShowWaitingDialog() }
//            .subscribeOn(Schedulers.io())
//            .observeOn(AndroidSchedulers.mainThread())
//            .map { responseBody: Response<ResponseBody?> ->
//                if (responseBody.isSuccessful && responseBody.code() == 207) {
//                    return@map responseBody.body()
//                } else {
//                    throw HttpException(responseBody)
//                }
//            }
//            .subscribe({ body: ResponseBody? -> saveWebDav(account) }) { throwable: Throwable? ->
//                fetchError(
//                    throwable
//                )
//            })
    }

    private fun saveWebDav(account: AccountsSqlData) {
//        account = false
//        mPreferenceTool.setDefault()
//        mPreferenceTool.portal = account.portal
//        mPreferenceTool.login = account.login
//        mPreferenceTool.password = account.password
//        mPreferenceTool.setVisitor(false)
//        mPreferenceTool.isNoPortal = false
//        account = true
//        viewState.onAccountLogin()
    }

    private fun login(account: CloudAccount?) {
        val token =
            AccountUtils.getToken(context, Account(account?.getAccountName(), context.getString(R.string.account_type)))
        if (token != null && token.isNotEmpty()) {
            viewState.onShowWaitingDialog()
            //onGetToken(account.token, account)
        } else {
            if (account?.isWebDav!!) {
//                viewState.onWebDavLogin(account)
            } else {
                account.login?.let { viewState.onSignIn(account.portal, it) }
            }
        }
    }

    //    @Override
    protected fun onErrorUser(response: Response<ResponseUser?>?) {
//        if (mClickedAccount != null) {
//            viewState.onSignIn(mClickedAccount!!.portal, mClickedAccount!!.login)
//        }
    }

    override fun onGetUser(user: User) {
        super.onGetUser(user)
        viewState.onAccountLogin()
    }

//    override var account: Boolean
//        get() = super.account
//        private set(isOnline) {
//            val account = mAccountSqlTool.getAccount(
//                mPreferenceTool.portal!!,
//                mPreferenceTool.login,
//                mPreferenceTool.socialProvider
//            )
//            if (account != null) {
//                account.isOnline = isOnline
//                mAccountSqlTool.setAccount(account)
//            }
//        }

    fun removeAccount() {
        if (contextAccount != null) {
            if (contextAccount!!.isOnline) {
                preferenceTool.setDefault()
            }
            CoroutineScope(Dispatchers.Default).launch {
                accountDao.deleteAccount(contextAccount!!)
                if (accountDao.getAccounts().isEmpty()) {
                    viewState.onEmptyList()
                } else {
                    withContext(Dispatchers.Main) {
                        viewState.removeItem(mContextPosition)
                    }
                }
            }
        }
    }

    fun logout() {
        if (contextAccount != null) {
            CoroutineScope(Dispatchers.Main).launch {
                AccountUtils.setPassword(
                    context,
                    Account(contextAccount?.getAccountName(), context.getString(R.string.account_type)),
                    ""
                )
                AccountUtils.setToken(
                    context,
                    Account(contextAccount?.getAccountName(), context.getString(R.string.account_type)),
                    ""
                )
                val logoutAccount = contextAccount?.copy(isOnline = false)
                logoutAccount?.let { accountDao.updateAccount(it) }
                logoutAccount?.let { viewState.onUpdateItem(it, mContextPosition) }
            }
        }
    }

    fun signIn() {
        if (contextAccount != null) {
            login(contextAccount!!)
        }
    }

    fun longClick(account: AccountsSqlData, position: Int) {
        //        mSelectedItems = ArrayList()
        //        isSelectionMode = true
        //        account.isSelection = true
        //        mSelectedItems.add(account)
        //        viewState.onSelectionMode()
        //        viewState.onActionBarTitle(mSelectedItems.size.toString())
    }

    fun onBackPressed(): Boolean {
        if (isSelectionMode) {
            isSelectionMode = false
            //            clearSelected()
            viewState.onDefaultState()
            return true
        }
        return false
    }

    //    private fun clearSelected() {
    //        for (account in mSelectedItems!!) {
    //            account.isSelection = false
    //        }
    //        mSelectedItems!!.clear()
    //        mSelectedItems = null
    //    }

    fun selectAll(itemList: MutableList<CloudAccount>) {
//            for (account in itemList) {
//                if (!account.isSelection && !account.isIdNull) {
//                    account.isSelection = true
//    //                mSelectedItems!!.add(account)
//                }
//            }
//            viewState.onActionBarTitle(mSelectedItems!!.size.toString())
//            viewState.onNotifyItems()
    }

    fun deleteAll() {
        if (mSelectedItems != null) {
            for (account in mSelectedItems!!) {
                if (account.isOnline) {
                    preferenceTool.setDefault()
                }
                //                mAccountSqlTool.delete(account)
            }
            onBackPressed()
            if (accountSqlTool.accounts.isEmpty()) {
                viewState.onEmptyList()
            } else {
                //                viewState.onSetAccounts(mAccountSqlTool.accounts)
            }
        }
    }

    fun deselectAll() {
        if (mSelectedItems != null) {
            for (account in mSelectedItems!!) {
//                account.isSelection = false
            }
            mSelectedItems!!.clear()
            viewState.onActionBarTitle(mSelectedItems!!.size.toString())
            viewState.onNotifyItems()
        }
    }

    //    private fun checkUpdate(accounts: List<CloudAccount>) {
    //        for (account in accounts) {
    //            if (!account.isWebDav && account.token != null && !account.token.isEmpty()) {
    //                mDisposable.add(getUser(account)
    //                    .subscribeOn(Schedulers.io())
    //                    .observeOn(AndroidSchedulers.mainThread())
    //                    .map { responseUser: ResponseUser ->
    //                        if (responseUser.response == null) {
    //                            return@map User()
    //                        } else {
    //                            return@map responseUser.response
    //                        }
    //                    }
    //                    .subscribe(
    //                        { user: User -> updateAccount(user, account) }
    //                    ) { throwable: Throwable -> Log.d(TAG, "checkUpdate: " + throwable.message) })
    //            }
    //        }
    //    }

    //    private fun updateAccount(user: User, account: AccountsSqlData) {
    //        if (!user.displayNameHtml.isEmpty() && account.name != user.displayNameHtml) {
    //            account.name = user.displayNameHtml
    //        }
    //        if (!user.avatar.isEmpty() && account.avatarUrl != user.avatar) {
    //            account.avatarUrl = user.avatar
    //        }
    //        mAccountSqlTool.setAccount(account)
    //        viewState.onNotifyItems()
    //    }
    //
    //    private fun getUser(account: AccountsSqlData): Observable<ResponseUser> {
    //        val token = account.token
    //        val retrofitTool = RetrofitTool(mContext)
    //        retrofitTool.isCiphers = account.isSslCiphers
    //        retrofitTool.isSslOn = account.isSslState
    //        return Observable.fromCallable {
    //            val response = retrofitTool.getApi(account.scheme + account.portal)
    //                .getUserInfo(token).execute()
    //            if (response.isSuccessful) {
    //                return@fromCallable response.body()
    //            } else {
    //                return@fromCallable ResponseUser()
    //            }
    //        }
    //    }
    //
    //
    //
    //    fun restoreAccount() {
    //        val accountsSqlData = mAccountSqlTool.accountOnline
    //        if (accountsSqlData != null) {
    //            mPreferenceTool.login = accountsSqlData.login
    //            mPreferenceTool.scheme = accountsSqlData.scheme
    //            mPreferenceTool.portal = accountsSqlData.portal
    //            mPreferenceTool.token = accountsSqlData.token
    //            mPreferenceTool.sslState = accountsSqlData.isSslState
    //            mPreferenceTool.sslCiphers = accountsSqlData.isSslCiphers
    //        }
    //    }

    fun accountClick(account: CloudAccount) {
        val token =
            AccountUtils.getToken(context, Account(account.getAccountName(), context.getString(R.string.account_type)))
        val portal = account.portal
        if (token != null && token.isNotEmpty() && portal != null && portal.isNotEmpty()) {
            disposable =
                service.getUserInfo(token).subscribe({ response ->
                    when (response) {
                        is LoginResponse.Success -> {
                            setAccount(account)
                            try {
                                viewState.onAccountLogin()
                            } catch (urlSyntaxMistake: UrlSyntaxMistake) {
                                urlSyntaxMistake.printStackTrace()
                            }
                        }
                        is LoginResponse.Error -> {
                            account.token = ""
                            account.login?.let { viewState.onSignIn(portal, it) }
                        }
                    }
                }, {
                    fetchError(it)
                })
        } else if (token != null && token.isEmpty()) {
            account.login?.let { viewState.onSignIn(portal, it) }
        } else {
            viewState.onError(context.getString(R.string.errors_sign_in_account_error))
        }
    }

    private fun setAccount(cloudAccount: CloudAccount) {
        CoroutineScope(Dispatchers.Default).launch {
            val account = accountDao.getAccountOnline()?.let { accountDao.getAccount(it.id)?.copy(isOnline = false) }
            account?.let { accountDao.updateAccount(it) }
            accountDao.updateAccount(cloudAccount.copy(isOnline = true))
        }
    }

    fun accountLongClick(account: CloudAccount) {
    }

//    fun selectAll(itemList: List<CloudAccount>?) {
//
//    }
//
//    fun deselectAll() {
//
//    }
//
//    fun deleteAll() {
//
//    }
}